### PR TITLE
Showing paramters in UaExpert

### PR DESCRIPTION
**Pull request by destogl**
Debug of TranslateBrowsePathsToNodeIds service. Still should be checked if it's implemented according to specification. Details in #34 

Current creation of new Arguments in the example-server is not so nice. We should think about a way to avid method *_guess_uatype* and enter type directly.

Also node.create_method should be extended so user can define a description.

Also mymethod is crashing the UaExpert... Don't merge for now.

### CODE REVIEW

**Comment by destogl** [comment]
The UaExpert is crashing because ValueRank paramters is per default set to 0 ([check here](https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/uaprotocol_auto.py#L784) and should be -1 if the data is scalar.

ValueRank is described in Part 3, Page 61, Table 23. Can we add this somewhere to be autogenerated as constant dictionary? Also automatic checking of this should be implemented.

File: opcua/address_space.py [hunk 1]
```diff
@@ -155,7 +156,7 @@ def _find_element_in_node(self, el, nodeid):
         nodedata = self._aspace[nodeid]
         for ref in nodedata.references:
             # FIXME: here we should check other arguments!!
-            if ref.BrowseName == el.TargetName:
+            if ref.BrowseName.Name == el.TargetName.Name:
                 return ref.NodeId
```
Comment (oroulet) [inline opcua/address_space.py]: that cannot be correct, they must be in same namespace

File: opcua/address_space.py [hunk 2]
```diff
@@ -139,6 +139,7 @@ def _translate_browsepath_to_nodeid(self, path):
             res.StatusCode = ua.StatusCode(ua.StatusCodes.BadNodeIdInvalid)
             return res
         current = path.StartingNode
+        # FIXME: Should be more according to the Specs: Part 4 Page 44
```
Comment (oroulet) [inline opcua/address_space.py]: what is missing? I do not remember there is any shortcut here

**Comment by destogl** [comment]
Maybe is nothing missing, but it was kind a strange so I wanted to check it again.

I send you tomorrow all logs, so you can see it.

**Comment by oroulet** [comment]
The UaExpert is crashing because ValueRank paramters is per default set to
0 (check here
<https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/uaprotocol_auto.py#L784>
and should be -1 if the data is scalar.

OK we need to fix this one

**Comment by oroulet** [comment]
You can override default value at the end of uaprotocol_hand.py

File: opcua/address_space.py (hunk 1 above)
Comment (destogl) [inline opcua/address_space.py]: see comment in #34

File: opcua/address_space.py (hunk 2 above)
Comment (destogl) [inline opcua/address_space.py]: Just checked it. It is good. Yesterday seamed strange... sorry

Even though you are not checking inverse references and subtypes... I don't know if this is important.

**Comment by destogl** [comment]
ValueRank is now corrected. I decided to define it at -2 per default because of description (Part 3, Page 61, Table 23):

```
Indicates whether the dataType is an array and how many dimensions the array has.
It may have the following values:
n > 1: the dataType is an array with the specified number of dimensions.
OneDimension (1): The dataType is an array with one dimension.
OneOrMoreDimensions (0): The dataType is an array with one or more dimensions.
Scalar (−1): The dataType is not an array.
Any (−2): The dataType can be a scalar or an array with any number of dimensions.
ScalarOrOneDimension (−3): The dataType can be a scalar or a one dimensional array.
NOTE All DataTypes are considered to be scalar, even if they have array-like semantics like ByteString and String.
```

**Comment by oroulet** [comment]
Is uaexpert happy now?

**Comment by oroulet** [comment]
Yes -2 looks like a good default. Just hope most clients support it....

**Comment by oroulet** [comment]
I add one more look. I could see the translatebrowsepath request asking for
a child with browsename 0:InputArguments. I then checked the prosys opc-ua
server (the prosys guys are very goos at following the opc-ua spec) and
there I could see that the browsename of inputarguments seems to always be
in namespace 0. event of method has another namespace. In the spec I read
"""
Servers may often choose to use the same namespace for the NodeId and the
BrowseName. However, if they want to provide a standard Property, its
BrowseName shall have the namespace of the standards body although the
namespace of the NodeId reflects something else, for example the local
server.
"""
that seems to confirm the InputArgument should use namespace 0.....and that
translatebrowsepath should compare namespace.
You need to modify the lines:
if inputs: create_property(method, qname.NamespaceIndex, "InputArguments",
[_vtype_to_argument(vtype) for vtype in inputs])
 if outputs: create_property(method, qname.NamespaceIndex,
"OutputArguments", [_vtype_to_argument(vtype) for vtype in outputs])

and write something like
ceate_property(method, GenerateNodeId(), QualifiedName(0,
"InputArguments"), [_vtype_to_argument(vtype) for vtype in inputs])
that shpuld make it

**Comment by destogl** [comment]
Now should be OK.
Should the FIXME from [here](https://github.com/FreeOpcUa/python-opcua/pull/35/files#diff-9bce3cb93ffe2484707e6e3b20c4881dL157) be removed?

I also corrected example-server mymethod didn't work.

File: example-server.py [hunk 3]
```diff
@@ -74,11 +77,32 @@ def multiply(parent, x, y):
...
+    inargy.ArrayDimensions = []
+    inargy.Description = ua.LocalizedText("Second number y")
+    outarg = ua.Argument()
+    outarg.Name = "Result"
+    outarg.DataType = node._guess_uatype(ua.Variant(None, ua.VariantType.Int64))
+    outarg.ValueRank = -1
+    outarg.ArrayDimensions = []
+    outarg.Description = ua.LocalizedText("Multiplication result")
+
+    multiply_node = myobj.add_method(idx, "multiply", multiply, [inargx, inargy], [outarg])
```
Comment (oroulet) [inline example-server.py]: I got several complains about the example file being too big, so I do not want to merge such a huge thing. Can you put it in another file? like for example-server-methods.py? Then we just leave the simplest method example in example-server.py?

File: example-server.py [hunk 4]
```diff
@@ -12,7 +12,7 @@ def embed():
         shell.interact()
 
 
-from opcua import ua, uamethod, Server, Event, ObjectIds
+from opcua import ua, uamethod, Server, Event, ObjectIds, node
```
Comment (oroulet) [inline example-server.py]: what is "node" with a small n? don't you get an error here?

Comment (destogl) [inline example-server.py]: It is the node.py file

**Pull request merged**